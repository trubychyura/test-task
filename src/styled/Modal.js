import styled from 'styled-components';
import { Col, Form, Image, Modal, Row } from 'react-bootstrap';

export const StyledModal = styled(Modal)`
  @media (max-width: 560px) {
    padding: 0   !important;
  }

  .modal-title {
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
  }

  .modal-dialog {
    max-width: 50%;
    @media (max-width: 1000px) {
      max-width: 70%;
    }
    @media (max-width: 580px) {
      max-width: 90%;
      margin: auto;
    }
  }
`;

export const StyledImage = styled(Image)`
  display: block;
  max-height: 300px;
  max-width: 100%;
  margin: auto;
`;

export const StyledForm = styled(Form)`
  width: 100%;
`;

export const StyledDescription = styled(Col)`
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

export const StyledComments = styled(Col)`
  padding: 15px;
`;

export const StyledComentHeader = styled(Row)`
  margin: 0;
  justify-content: space-between;
`;
