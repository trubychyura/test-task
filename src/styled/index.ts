export const theme = {
  color: {
    gold: 'gold',
    silver: 'silver',
  },
  media: {
    md: '580px',
    lg: '1000px',
  },
};
