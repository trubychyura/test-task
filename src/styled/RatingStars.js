import styled from 'styled-components';
import { Star } from '@styled-icons/boxicons-solid/Star';
import { Form } from 'react-bootstrap';

export const StyledStar = styled(Star)`
  cursor: pointer;
  transition: color 200ms;

  ${({ value, rating, hover }) =>
    value <= (hover || rating) ? `color:  gold;` : `color: silver;`}
`;

export const StyledRadio = styled(Form.Check)`
  display: none;
`;

export const StyledContainer = styled(Form.Group)`
  display: flex;
`;
