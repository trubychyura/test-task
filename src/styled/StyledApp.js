import styled from 'styled-components';
import { Container, Table } from 'react-bootstrap';

export const StyledContainer = styled(Container)`
  padding: 20px 10px;
`;

export const StyledTable = styled(Table)`
  text-align: center;
  th {
    font-size: 40px;
  }

  tbody tr {
    cursor: pointer;
  }
`;
