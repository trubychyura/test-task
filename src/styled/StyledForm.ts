import styled from 'styled-components';
import { Alert, Button, Form } from 'react-bootstrap';

export const StyledForm = styled(Form)`
  width: 100%;
`;

export const StyledAlert = styled(Alert)`
  font-size: 13px;
  margin: 5px 0;
  padding: 5px;
`;

export const StyledButton = styled(Button)`
  padding-left: 30px;
  padding-right: 30px;
`;
