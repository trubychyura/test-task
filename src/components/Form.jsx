import { Formik } from 'formik';
import { RatingStars } from '.';
import { StyledForm, StyledButton, StyledAlert } from '../styled/StyledForm';
import * as yup from 'yup';

const Form = () => (
  <Formik
    initialValues={{
      comment: '',
      rating: 0,
    }}
    onSubmit={(props) => {
      console.log(props);
    }}
    validationSchema={yup.object({
      comment: yup.string('Must be a string').required('Comment is required!'),
      rating: yup
        .number()
        .moreThan(0, "Product's rate is required!")
        .required("Product's rate is required!"),
    })}
  >
    {(formik) => (
      <StyledForm onSubmit={formik.handleSubmit}>
        <StyledForm.Group>
          <StyledForm.Label>Products's rate:</StyledForm.Label>
          <RatingStars name='rating' />

          {formik.errors.rating && formik.touched.rating && (
            <StyledAlert variant='danger'>{formik.errors.rating}</StyledAlert>
          )}
        </StyledForm.Group>
        <StyledForm.Group controlId='formComment'>
          <StyledForm.Label>Your comment:</StyledForm.Label>
          <StyledForm.Control
            as='textarea'
            value={formik.values.comment}
            placeholder='Enter comment'
            name='comment'
            onChange={formik.handleChange}
          />

          {formik.errors.comment && formik.touched.comment && (
            <StyledAlert variant='danger'>{formik.errors.comment}</StyledAlert>
          )}
        </StyledForm.Group>
        <StyledButton variant='primary' type='submit'>
          Send
        </StyledButton>
      </StyledForm>
    )}
  </Formik>
);

export default Form;
