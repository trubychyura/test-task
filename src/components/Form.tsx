import { Formik } from 'formik';
import { RatingStars } from '.';
import { StyledForm, StyledButton, StyledAlert } from '../styled/StyledForm';
import * as yup from 'yup';
import { connect } from 'react-redux';
import { handleSubmit } from '../redux/products';

type FormProps = {
  id: string;
  handleSubmit: (id: string, data: any) => void;
};

const Form = ({ id, handleSubmit }: FormProps) => (
  <Formik
    initialValues={{
      comment: '',
      rating: 0,
    }}
    onSubmit={(data, { resetForm }) => {
      handleSubmit(id, { text: data.comment, rating: data.rating });
      resetForm();
    }}
    validationSchema={yup.object({
      comment: yup.string().required('Comment is required!'),
      rating: yup
        .number()
        .moreThan(0, "Product's rate is required!")
        .required("Product's rate is required!"),
    })}
  >
    {(formik) => (
      <StyledForm onSubmit={formik.handleSubmit}>
        <StyledForm.Group>
          <StyledForm.Label>Products's rate:</StyledForm.Label>
          {/* <RatingStars name='rating' /> */}
          {formik.errors.rating && formik.touched.rating && (
            <StyledAlert variant='danger'>{formik.errors.rating}</StyledAlert>
          )}
        </StyledForm.Group>
        <StyledForm.Group controlId='formComment'>
          <StyledForm.Label>Your comment:</StyledForm.Label>
          <StyledForm.Control
            as='textarea'
            value={formik.values.comment}
            placeholder='Enter comment'
            name='comment'
            onChange={formik.handleChange}
          />
          {formik.errors.comment && formik.touched.comment && (
            <StyledAlert variant='danger'>{formik.errors.comment}</StyledAlert>
          )}
        </StyledForm.Group>
        <StyledButton variant='primary' type='submit'>
          Send
        </StyledButton>
      </StyledForm>
    )}
  </Formik>
);

export default connect(null, { handleSubmit })(Form);
