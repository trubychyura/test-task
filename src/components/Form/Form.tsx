import { FC } from 'react';
import { connect } from 'react-redux';
import { Formik, FormikProps } from 'formik';
import * as yup from 'yup';

import { RatingStars } from '..';

import { handleSubmit } from '../../redux/ducks/products';
import { FormProps, FormikValues } from '../../types';
import { StyledForm, StyledButton, StyledAlert } from '../../styled/StyledForm';

const initialValues: FormikValues = {
  comment: '',
  rating: 0,
};

export const Form: FC<FormProps> = ({ id, handleSubmit }) => (
  <Formik
    validateOnBlur={true}
    initialValues={initialValues}
    onSubmit={(values, { resetForm }) => {
      handleSubmit(id, { text: values.comment, rating: values.rating });
      resetForm();
    }}
    validationSchema={yup.object({
      comment: yup.string().required('Comment is required!'),
      rating: yup
        .number()
        .moreThan(0, "Product's rate is required!")
        .required("Product's rate is required!"),
    })}
  >
    {(formik: FormikProps<FormikValues>) => (
      <StyledForm onSubmit={formik.handleSubmit} data-testid='form'>
        <StyledForm.Group>
          <StyledForm.Label>Products's rate:</StyledForm.Label>
          <RatingStars name='rating' />
          {formik.errors.rating && formik.touched.rating && (
            <StyledAlert variant='danger'>{formik.errors.rating}</StyledAlert>
          )}
        </StyledForm.Group>
        <StyledForm.Group controlId='formComment'>
          <StyledForm.Label>Your comment:</StyledForm.Label>
          <StyledForm.Control
            value={formik.values.comment}
            as='textarea'
            placeholder='Enter comment'
            name='comment'
            onChange={formik.handleChange}
          />
          {formik.errors.comment && formik.touched.comment && (
            <StyledAlert variant='danger' data-testid='error'>
              {formik.errors.comment}
            </StyledAlert>
          )}
        </StyledForm.Group>
        <StyledButton variant='primary' type='submit'>
          Send
        </StyledButton>
      </StyledForm>
    )}
  </Formik>
);

export default connect(null, { handleSubmit })(Form);
