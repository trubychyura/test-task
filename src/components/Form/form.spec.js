import { shallow } from 'enzyme';

import { Form } from './Form';
import { Form as BootstrapForm } from 'react-bootstrap';
import { StyledForm } from '../../styled/StyledForm';

import { theme } from '../../styled';

const setUp = (props) => shallow(<Form {...props} />);

describe('Form component', () => {
  let component;
  beforeEach(() => {
    component = setUp();
  });

  it('should render component', () => {
    expect(component.dive()).toMatchSnapshot();
  });

  it('should update Form.Control when it is changed', () => {
    // console.log(component.dive().find(StyledForm.Control).debug());
    // const input = component.dive().find(StyledForm.Control);
    // console.log(input.debug());
    // input.prop('onChange', {
    //   target: { name: 'comment', value: 'new comment' },
    // });
    // console.log(input.debug());
    // input.simulate('change', {
    //   // you must add this next line as (Formik calls e.persist() internally)
    //   persist: () => {},
    //   // simulate changing e.target.name and e.target.value
    //   target: {
    //     name: 'comment',
    //     value: 'ian',
    //   },
    // });
    // const newValue = input.props();
    // console.log(input.props());
    // expect(newValue.value).toEqual('new comment');
  });
});
