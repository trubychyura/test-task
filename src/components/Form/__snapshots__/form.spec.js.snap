// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should render component 1`] = `
<ContextProvider
  value={
    Object {
      "dirty": false,
      "errors": Object {},
      "getFieldHelpers": [Function],
      "getFieldMeta": [Function],
      "getFieldProps": [Function],
      "handleBlur": [Function],
      "handleChange": [Function],
      "handleReset": [Function],
      "handleSubmit": [Function],
      "initialErrors": Object {},
      "initialStatus": undefined,
      "initialTouched": Object {},
      "initialValues": Object {
        "comment": "",
        "rating": 0,
      },
      "isSubmitting": false,
      "isValid": true,
      "isValidating": false,
      "registerField": [Function],
      "resetForm": [Function],
      "setErrors": [Function],
      "setFieldError": [Function],
      "setFieldTouched": [Function],
      "setFieldValue": [Function],
      "setFormikState": [Function],
      "setStatus": [Function],
      "setSubmitting": [Function],
      "setTouched": [Function],
      "setValues": [Function],
      "status": undefined,
      "submitCount": 0,
      "submitForm": [Function],
      "touched": Object {},
      "unregisterField": [Function],
      "validateField": [Function],
      "validateForm": [Function],
      "validateOnBlur": true,
      "validateOnChange": true,
      "validateOnMount": false,
      "values": Object {
        "comment": "",
        "rating": 0,
      },
    }
  }
>
  <Styled(Form)
    onSubmit={[Function]}
  >
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Products's rate:
      </FormLabel>
      <RatingStars
        name="rating"
      />
    </FormGroup>
    <FormGroup
      controlId="formComment"
    >
      <FormLabel
        column={false}
        srOnly={false}
      >
        Your comment:
      </FormLabel>
      <FormControl
        as="textarea"
        name="comment"
        onChange={[Function]}
        placeholder="Enter comment"
        value=""
      />
    </FormGroup>
    <Styled(Button)
      type="submit"
      variant="primary"
    >
      Send
    </Styled(Button)>
  </Styled(Form)>
</ContextProvider>
`;
